src/sqlfluff/rules/L053.py

```diff
@@ -0,0 +1,75 @@
+"""Implementation of Rule L053."""
+from typing import Optional
+
+from sqlfluff.core.rules.base import BaseRule, LintResult, LintFix, RuleContext
+from sqlfluff.core.rules.doc_decorators import document_fix_compatible
+
+
+@document_fix_compatible
+class Rule_L053(BaseRule):
+    """Top-level statements should not be wrapped in brackets.
+
+    | **Anti-pattern**
+    | A top-level statement is wrapped in brackets.
+
+    .. code-block:: sql
+       :force:
+
+        (SELECT
+            foo
+        FROM bar)
+
+        -- This also applies to statements containing a sub-query.
+
+        (SELECT
+            foo
+        FROM (SELECT * FROM bar))
+
+    | **Best practice**
+    | Don't wrap top-level statements in brackets.
+
+    .. code-block:: sql
+       :force:
+
+        SELECT
+            foo
+        FROM bar
+
+        -- Likewise for statements containing a sub-query.
+
+        SELECT
+            foo
+        FROM (SELECT * FROM bar)
+    """
+
+    def _eval(self, context: RuleContext) -> Optional[LintResult]:
+        """Top-level statements should not be wrapped in brackets."""
+        # We only care about bracketed segements that are direct
+        # descendants of a top-level statement segment.
+        if context.dialect.name == "tsql":
+            top_level_parent_stack_types = ["file", "batch", "statement"]
```

Comment:
I wonder if we could get this kind of information from the dialect, rather than storing it on the rules. Rules that are dialect specific will be harder to maintain.

Reply:
Is there another rule where we do something similar?

Reply:
Do we even need this to be t-sql specific? It seems "batch" is only use din t-sql so why not just include this as the default `top_level_parent_stack_types` and remove this if statement completely?

Reply:
I guess because I'm comparing this to the parent stack for an exact match to prove I'm at the top level

Reply:
I could maybe change the filter logic to something like:
```python
if not (
    context.segment.is_type("bracketed")
    and [segment.type for segment in context.parent_stack if segment.type != "batch"]
    == ["file", "statement"]
):
    return None
```
so effectively just ignoring the batch.

It's not perfect as it still references batch (and therefore tsql) and potentially less clearly than before 🤷 Not sure of another way to check that the brackets are at the top-level?

Reply:
Could you do what I suggested here: https://github.com/sqlfluff/sqlfluff/pull/1916#discussion_r751810281

So, for example this:

```sql
(SELECT * FROM table1)
UNION
(SELECT * FROM table2)
```
Is this:

```
file
|_statement
   |_set_expression
      |_bracketed
      |_set_operator
      |_bracketed
```

Therefore as it has a `set_expression` in it's parent tree back to `file` so it's not a top level bracket.

Whereas this:

```sql

(SELECT * FROM table1)
```
Is this:

```
file
|_newline
|_statement
   |_bracketed
```

So from `bracketed` back to `file` is `bracketed->statement->file` so it only contains the allowed types. So it is a top level statement.

Note I deliberately included a newline to show it's a parallel tree and so not part of the statement.

WDYT?

Reply:
or could do `{segment.type for segment in context.parent_stack}.is_subset({"file", "batch", "statement"})` which serves the same purpose and is a bit more concise (If we don't care about ordering/counts)

Reply:
Exactly. Can't see why that wouldn't work, can you?

Reply:
@tunetheweb I think that makes sense, that's how the code works already. I guess there's a question around whether we should also be able remove brackets around those statements in the set expression?

Reply:
> I guess there's a question around whether we should also be able remove brackets around those statements in the set expression?

I don't think so. It's not a top level segment - it's two bracketed sub segments with an unbracketed `UNION` in between.

Reply:
In fact I’m not even sure we should count `batch` in this. Is that for this use case?

```sql
BEGIN
  (SELECT 1)
END
```

I don’t think the rule should flag that as not top level either.

I think keep it simple just to File and Statement and that’s sufficient.

Reply:
> Exactly. Can't see why that wouldn't work, can you?

I think if we know that we're looking for 'file->(batch)->statement->brackets' exactly then we might as well be sure that our parent chain matches that.

The set method would concern me if we could match 'file->(batch)->statement->statement->statement->brackets' which, although I can't visualise what that is, probably wouldn't be top level

Reply:
> Exactly. Can't see why that wouldn't work, can you?

I think if we know that we're looking for 'file->(batch)->statement->brackets' exactly then we might as well be sure that our parent chain matches that.

The set method would concern me if we could match 'file->(batch)->statement->statement->statement->brackets' which, although I can't visualise what that is, probably wouldn't be top level

Reply:
@tunetheweb actually the begin/end case you mention above has parents like 'file->(batch)->begin_end_block->statement->brackets' so wouldn't be touched by this rule, which is what we expect

![image](https://user-images.githubusercontent.com/80432516/142404930-b399c32a-066c-4fa1-9df0-2cb30891ccfd.png)

Reply:
Whereas 'file->(batch)->statement->brackets' covers the top level case we want to catch:
![image](https://user-images.githubusercontent.com/80432516/142405271-38e70238-cb2e-41e7-9a79-11889d8b19a5.png)

Reply:
Ah OK for T-SQL everything is in a "batch". Gotcha.

OK then I agree 'file->(batch)->statement->brackets' is the way to go, and no explicit override for T-SQL.

Reply:
perfect, I've updated the logic to remove the tsql reference. I've also added extra unit tests to validate that we aren't getting the false positives discussed 👍